File: packages/elixir/lib/codex_sdk/app_server.ex
defmodule CodexSdk.AppServer do
  @moduledoc """
  Low-level stdio client for a local Codex app-server subprocess.
  """

  use GenServer

  defmodule JsonRpcError do
    @moduledoc "JSON-RPC error response."
    defexception [:code, :data, message: "JSON-RPC error"]
  end

  defmodule ClosedError do
    @moduledoc "Raised when the app-server process exits while requests are pending."
    defexception [:code, message: "app-server closed"]
  end

  defstruct port: nil,
            owner: nil,
            decoder: CodexSdk.LineFraming.new(),
            router: CodexSdk.MessageRouter.new(),
            pending: %{},
            counter: 0,
            closed?: false,
            exit_code: nil

  @type event ::
          {:notification, map()}
          | {:server_request, String.t() | number(), map()}
          | {:malformed, String.t()}
          | {:unknown, map()}
          | {:exit, integer()}

  @spec start_link(keyword()) :: GenServer.on_start()
  def start_link(opts \\ []) do
    opts = Keyword.put_new(opts, :owner, self())
    GenServer.start_link(__MODULE__, opts)
  end

  @spec request(GenServer.server(), String.t(), term(), timeout()) ::
          {:ok, term()} | {:error, JsonRpcError.t() | ClosedError.t()}
  def request(server, method, params \\ nil, timeout \\ 5_000) do
    GenServer.call(server, {:request, method, params}, timeout)
  end

  @spec notify(GenServer.server(), String.t(), term()) :: :ok | {:error, ClosedError.t()}
  def notify(server, method, params \\ nil) do
    GenServer.call(server, {:notify, method, params})
  end

  @spec respond(GenServer.server(), String.t() | number(), term()) ::
          :ok | {:error, ClosedError.t()}
  def respond(server, id, result) do
    GenServer.call(server, {:respond, id, result})
  end

  @spec respond_error(GenServer.server(), String.t() | number(), map()) ::
          :ok | {:error, ClosedError.t()}
  def respond_error(server, id, error) do
    GenServer.call(server, {:respond_error, id, error})
  end

  @spec stop(GenServer.server()) :: :ok
  def stop(server) do
    GenServer.call(server, :stop)
  catch
    :exit, _reason -> :ok
  end

  @impl true
  def init(opts) do
    command = Keyword.get(opts, :command, "codex")
    args = Keyword.get(opts, :args, ["app-server", "--listen", "stdio://"])
    owner = Keyword.fetch!(opts, :owner)

    executable = System.find_executable(command) || command

    port_opts =
      [:binary, :exit_status, args: args]
      |> maybe_put_cd(Keyword.get(opts, :cwd))

    port = Port.open({:spawn_executable, executable}, port_opts)

    {:ok, %__MODULE__{port: port, owner: owner}}
  end

  @impl true
  def handle_call({:request, _method, _params}, _from, %__MODULE__{closed?: true} = state) do
    {:reply, {:error, closed_error(state.exit_code)}, state}
  end

  def handle_call({:request, method, params}, from, %__MODULE__{} = state) do
    id = "req-#{state.counter + 1}"

    message =
      %{"id" => id, "method" => method}
      |> maybe_put_params(params)

    case send_message(state, message) do
      :ok ->
        state = %{
          state
          | counter: state.counter + 1,
            router: CodexSdk.MessageRouter.expect_response(state.router, id),
            pending: Map.put(state.pending, id, from)
        }

        {:noreply, state}

      {:error, error} ->
        {:reply, {:error, error}, state}
    end
  end

  def handle_call({:notify, method, params}, _from, %__MODULE__{} = state) do
    message =
      %{"method" => method}
      |> maybe_put_params(params)

    {:reply, send_message(state, message), state}
  end

  def handle_call({:respond, id, result}, _from, %__MODULE__{} = state) do
    {:reply, send_message(state, %{"id" => id, "result" => result}), state}
  end

  def handle_call({:respond_error, id, error}, _from, %__MODULE__{} = state) do
    {:reply, send_message(state, %{"id" => id, "error" => error}), state}
  end

  def handle_call(:stop, _from, %__MODULE__{} = state) do
    state =
      if state.closed? do
        state
      else
        Port.close(state.port)
        %{state | closed?: true}
      end

    {:reply, :ok, state}
  end

  @impl true
  def handle_info({port, {:data, data}}, %__MODULE__{port: port} = state) do
    {decoder, events} = CodexSdk.LineFraming.feed(state.decoder, data)

    state =
      events
      |> Enum.reduce(%{state | decoder: decoder}, &handle_event/2)

    {:noreply, state}
  end

  def handle_info({port, {:exit_status, status}}, %__MODULE__{port: port} = state) do
    state =
      if state.closed? do
        state
      else
        Enum.each(state.pending, fn {_id, from} ->
          GenServer.reply(from, {:error, closed_error(status)})
        end)

        send_event(state, {:exit, status})

        %{state | closed?: true, exit_code: status, pending: %{}}
      end

    {:noreply, state}
  end

  defp handle_event({:malformed, raw}, %__MODULE__{} = state) do
    send_event(state, {:malformed, raw})
    state
  end

  defp handle_event({:message, message}, %__MODULE__{} = state) do
    {router, routed} = CodexSdk.MessageRouter.route(state.router, message)

    state
    |> Map.put(:router, router)
    |> handle_routed(routed)
  end

  defp handle_routed(%__MODULE__{} = state, {:response, id, message}) do
    reply_pending(state, id, {:ok, message["result"]})
  end

  defp handle_routed(%__MODULE__{} = state, {:error_response, id, message}) do
    reply_pending(state, id, {:error, json_rpc_error(message["error"])})
  end

  defp handle_routed(%__MODULE__{} = state, {:server_request, id, message}) do
    send_event(state, {:server_request, id, message})
    state
  end

  defp handle_routed(%__MODULE__{} = state, {:notification, message}) do
    send_event(state, {:notification, message})
    state
  end

  defp handle_routed(%__MODULE__{} = state, {:orphan_response, _id, message}) do
    send_event(state, {:unknown, message})
    state
  end

  defp handle_routed(%__MODULE__{} = state, {:unknown, message}) do
    send_event(state, {:unknown, message})
    state
  end

  defp reply_pending(%__MODULE__{} = state, id, reply) do
    case Map.pop(state.pending, id) do
      {nil, pending} ->
        %{state | pending: pending}

      {from, pending} ->
        GenServer.reply(from, reply)
        %{state | pending: pending}
    end
  end

  defp send_message(%__MODULE__{closed?: true, exit_code: code}, _message) do
    {:error, closed_error(code)}
  end

  defp send_message(%__MODULE__{} = state, message) do
    if Port.command(state.port, CodexSdk.JsonRpc.encode(message)) do
      :ok
    else
      {:error, closed_error(state.exit_code)}
    end
  end

  defp maybe_put_params(message, nil), do: message
  defp maybe_put_params(message, params), do: Map.put(message, "params", params)

  defp maybe_put_cd(opts, nil), do: opts
  defp maybe_put_cd(opts, cwd), do: [{:cd, cwd} | opts]

  defp send_event(%__MODULE__{} = state, event) do
    send(state.owner, {:usetemi_codex_sdk_app_server_event, self(), event})
  end

  defp json_rpc_error(error) when is_map(error) do
    %JsonRpcError{
      code: Map.get(error, "code", -32_000),
      message: Map.get(error, "message", "JSON-RPC error"),
      data: Map.get(error, "data")
    }
  end

  defp json_rpc_error(error) do
    %JsonRpcError{code: -32_000, message: "JSON-RPC error", data: error}
  end

  defp closed_error(code) do
    %ClosedError{code: code, message: "app-server closed with code #{inspect(code)}"}
  end
end

File: packages/go/app_server.go
package codexsdk

import (
	"bufio"
	"context"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sync"
)

// AppServerOptions configures a local Codex app-server subprocess.
type AppServerOptions struct {
	Command string
	Args    []string
	Dir     string
	Env     []string
}

// JSONRPCErrorBody is the JSON-RPC error object returned by the app-server.
type JSONRPCErrorBody struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
	Data    any    `json:"data,omitzero"`
}

// JSONRPCResponseError wraps a JSON-RPC error response.
type JSONRPCResponseError struct {
	Code    int
	Message string
	Data    any
}

func (err *JSONRPCResponseError) Error() string {
	return err.Message
}

// AppServerClosedError is returned when the app-server exits while requests are pending.
type AppServerClosedError struct {
	Code *int
}

func (err *AppServerClosedError) Error() string {
	if err.Code == nil {
		return "app-server closed"
	}

	return fmt.Sprintf("app-server closed with code %d", *err.Code)
}

// AppServerEventType identifies an app-server event.
type AppServerEventType string

const (
	AppServerEventNotification  AppServerEventType = "notification"
	AppServerEventServerRequest AppServerEventType = "serverRequest"
	AppServerEventMalformed     AppServerEventType = "malformed"
	AppServerEventUnknown       AppServerEventType = "unknown"
	AppServerEventExit          AppServerEventType = "exit"
)

// AppServerEvent is emitted for server notifications, server requests, malformed output, and exits.
type AppServerEvent struct {
	Type    AppServerEventType
	ID      any
	Message JSONRPCMessage
	Raw     string
	Code    *int
}

type responseOutcome struct {
	result any
	err    error
}

// AppServerClient manages a local stdio Codex app-server subprocess.
type AppServerClient struct {
	cmd    *exec.Cmd
	stdin  io.WriteCloser
	events chan AppServerEvent

	decoder JSONLineDecoder
	router  *MessageRouter

	mutex          sync.Mutex
	pending        map[responseID]chan responseOutcome
	requestCounter int
	closed         bool

	waitDone chan struct{}
	waitErr  error
}

// StartAppServer starts a Codex app-server subprocess.
func StartAppServer(ctx context.Context, options AppServerOptions) (*AppServerClient, error) {
	command := options.Command
	if command == "" {
		command = "codex"
	}
	args := options.Args
	if args == nil {
		args = []string{"app-server", "--listen", "stdio://"}
	}

	cmd := exec.CommandContext(ctx, command, args...)
	cmd.Dir = options.Dir
	if options.Env != nil {
		cmd.Env = append(os.Environ(), options.Env...)
	}

	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, fmt.Errorf("open app-server stdin: %w", err)
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("open app-server stdout: %w", err)
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return nil, fmt.Errorf("open app-server stderr: %w", err)
	}

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("start app-server: %w", err)
	}

	client := &AppServerClient{
		cmd:      cmd,
		stdin:    stdin,
		events:   make(chan AppServerEvent, 32),
		router:   NewMessageRouter(),
		pending:  make(map[responseID]chan responseOutcome),
		waitDone: make(chan struct{}),
	}

	go io.Copy(io.Discard, stderr)
	go client.readStdout(stdout)
	go client.wait()

	return client, nil
}

// Request sends a JSON-RPC request and waits for its response.
func (client *AppServerClient) Request(ctx context.Context, method string, params any) (any, error) {
	id := client.nextRequestID()
	key, _ := normalizeResponseID(id)
	responses := make(chan responseOutcome, 1)

	client.mutex.Lock()
	client.pending[key] = responses
	client.router.ExpectResponse(id)
	client.mutex.Unlock()

	if err := client.write(JSONRPCMessage{"id": id, "method": method, "params": params}); err != nil {
		client.removePending(key)
		return nil, err
	}

	select {
	case response := <-responses:
		return response.result, response.err
	case <-ctx.Done():
		client.removePending(key)
		return nil, context.Cause(ctx)
	}
}

// Notify sends a JSON-RPC notification.
func (client *AppServerClient) Notify(method string, params any) error {
	return client.write(JSONRPCMessage{"method": method, "params": params})
}

// Respond sends a successful response to a server-initiated request.
func (client *AppServerClient) Respond(id any, result any) error {
	return client.write(JSONRPCMessage{"id": id, "result": result})
}

// RespondError sends an error response to a server-initiated request.
func (client *AppServerClient) RespondError(id any, errorBody JSONRPCErrorBody) error {
	return client.write(JSONRPCMessage{"id": id, "error": errorBody})
}

// Events returns app-server events.
func (client *AppServerClient) Events() <-chan AppServerEvent {
	return client.events
}

// Close stops the app-server subprocess. It is safe to call more than once.
func (client *AppServerClient) Close() error {
	client.mutex.Lock()
	alreadyClosed := client.closed
	client.mutex.Unlock()
	if alreadyClosed {
		return nil
	}

	_ = client.stdin.Close()
	if client.cmd.Process != nil {
		_ = client.cmd.Process.Kill()
	}
	<-client.waitDone
	return client.waitErr
}

func (client *AppServerClient) nextRequestID() string {
	client.mutex.Lock()
	defer client.mutex.Unlock()

	client.requestCounter++
	return fmt.Sprintf("req-%d", client.requestCounter)
}

func (client *AppServerClient) write(message JSONRPCMessage) error {
	client.mutex.Lock()
	defer client.mutex.Unlock()

	if client.closed {
		return &AppServerClosedError{}
	}

	encoded, err := EncodeJSONRPCMessage(message)
	if err != nil {
		return err
	}
	if _, err := client.stdin.Write(encoded); err != nil {
		return fmt.Errorf("write app-server message: %w", err)
	}

	return nil
}

func (client *AppServerClient) readStdout(stdout io.Reader) {
	reader := bufio.NewReader(stdout)
	for {
		chunk, err := reader.ReadBytes('\n')
		if len(chunk) > 0 {
			decoded := client.decoder.Feed(chunk)
			for _, malformed := range decoded.Malformed {
				client.events <- AppServerEvent{Type: AppServerEventMalformed, Raw: malformed.Raw}
			}
			for _, message := range decoded.Messages {
				client.handleMessage(message)
			}
		}
		if err != nil {
			return
		}
	}
}

func (client *AppServerClient) handleMessage(message JSONRPCMessage) {
	routed := client.router.Route(message)
	switch routed.Type {
	case RouteResponse:
		client.resolvePending(routed.ID, responseOutcome{result: message["result"]})
	case RouteErrorResponse:
		client.resolvePending(routed.ID, responseOutcome{err: parseJSONRPCResponseError(message["error"])})
	case RouteServerRequest:
		client.events <- AppServerEvent{Type: AppServerEventServerRequest, ID: routed.ID, Message: message}
	case RouteNotification:
		client.events <- AppServerEvent{Type: AppServerEventNotification, Message: message}
	case RouteUnknown, RouteOrphanResponse:
		client.events <- AppServerEvent{Type: AppServerEventUnknown, Message: message}
	}
}

func (client *AppServerClient) resolvePending(id any, outcome responseOutcome) {
	key, ok := normalizeResponseID(id)
	if !ok {
		return
	}

	client.mutex.Lock()
	responses := client.pending[key]
	delete(client.pending, key)
	client.mutex.Unlock()

	if responses != nil {
		responses <- outcome
	}
}

func (client *AppServerClient) removePending(key responseID) {
	client.mutex.Lock()
	delete(client.pending, key)
	client.mutex.Unlock()
}

func (client *AppServerClient) wait() {
	err := client.cmd.Wait()
	code := 0
	if client.cmd.ProcessState != nil {
		code = client.cmd.ProcessState.ExitCode()
	}
	if err != nil && errors.Is(err, os.ErrProcessDone) {
		err = nil
	}

	client.mutex.Lock()
	if client.closed {
		client.mutex.Unlock()
		close(client.waitDone)
		return
	}
	client.closed = true
	pending := make([]chan responseOutcome, 0, len(client.pending))
	for _, responses := range client.pending {
		pending = append(pending, responses)
	}
	clear(client.pending)
	client.waitErr = err
	client.mutex.Unlock()

	closedError := &AppServerClosedError{Code: new(code)}
	for _, responses := range pending {
		responses <- responseOutcome{err: closedError}
	}
	client.events <- AppServerEvent{Type: AppServerEventExit, Code: new(code)}
	close(client.waitDone)
}

func parseJSONRPCResponseError(raw any) *JSONRPCResponseError {
	errorBody, ok := raw.(map[string]any)
	if !ok {
		return &JSONRPCResponseError{Code: -32000, Message: "JSON-RPC error"}
	}

	code := -32000
	if value, ok := errorBody["code"].(float64); ok {
		code = int(value)
	}
	message := "JSON-RPC error"
	if value, ok := errorBody["message"].(string); ok {
		message = value
	}

	return &JSONRPCResponseError{
		Code:    code,
		Message: message,
		Data:    errorBody["data"],
	}
}
